### PR TITLE
[Performance][SFA] Generalize multi-FIA prefill to heterogeneous requests

### PR DESCRIPTION
**Pull request by Ruiqiu-Zheng**
This is a stacked follow-up depending on #16091. It subsumes the exact three-request Multi-FIA implementation from #16267 while retaining that case as regression coverage, and generalizes the production path to heterogeneous requests.

Scope:
- Generalizes multi-FIA prefill to heterogeneous request-local activation, keeping the existing default-off and fail-closed behavior.
- Limits the incremental parent-to-generic change to the SFA v1 implementation and unit coverage.
- Does not include DCP, decode, LightningIndexer sharding, or unrelated mechanisms.

Evidence:
- Historical generic run matched TP16 primary p50 TTFT 3628.501 -> 3660.905 ms (+0.893%, slightly worse).
- Historical generic run p90 TTFT improved by 9.934%.
- Historical generic run batch makespan p50 5423.315 -> 4927.731 ms (-9.138%).
- Correctness: PASS.
- Coverage evidence: 57 eligible layers and request-local 2/3 segment activation.

Claim ceiling:
- This PR makes no median-TTFT-gain claim.
- This PR makes no production claim.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/attention/sfa_v1.py [hunk 1]
```diff
@@ -1124,6 +1202,515 @@ def _execute_sparse_flash_attention_process(
...
+
+                tail_output = None
+                if tail_spans:
+                    tail_q = pack_multifia_rows(ql_nope, tail_spans)
+                    tail_rope = pack_multifia_rows(q_pe, tail_spans)
+                    tail_indices = pack_multifia_rows(topk_indices, tail_spans)
+                    tail_metadata = copy(attn_metadata)
+                    tail_metadata.block_table = attn_metadata.block_table[list(tail_requests)]
+                    tail_metadata.cum_query_lens = cum_query_lens.new_tensor(list(tail_query_ends))
+                    tail_metadata.seq_lens = seq_lens.new_tensor(list(tail_kv_lengths))
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/sfa_v1.py:1770]: Validated against current HEAD c14517a. On this multi-FIA tail path, the request-local `cum_query_lens`/`seq_lens` tensors are passed explicitly to `_execute_sparse_flash_attention_process`, and the downstream `DeviceOperator.execute_sparse_flash_attention_process` consumes those explicit length tensors plus the explicit block table; it does not read `cum_query_lens_cpu`/`seq_lens_cpu`. This copied metadata object is not reused after the tail call/row restoration, so the inherited CPU fields are inert on this path. I’m keeping this unchanged to avoid metadata-only churn with no observable consumer.

File: vllm_ascend/attention/sfa_v1.py [hunk 2]
```diff
@@ -1124,6 +1202,515 @@ def _execute_sparse_flash_attention_process(
...
+                return tensor[start:end]
+            return torch.cat([tensor[start:end] for start, end in spans], dim=0)
+
+        dense_q = pack_rows(ql_nope, dense_spans)
+        dense_rope = pack_rows(q_pe, dense_spans)
+        dense_metadata = copy(attn_metadata)
+        dense_metadata.attn_state = AscendAttentionState.PrefillCacheHit
+        dense_metadata.block_table = attn_metadata.block_table[list(dense_requests)]
+        dense_metadata.cum_query_lens_cpu = cum_query_lens_cpu.new_tensor(list(dense_query_ends))
+        dense_metadata.seq_lens_cpu = seq_lens_cpu.new_tensor(list(dense_kv_lengths))
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/sfa_v1.py:1886]: Validated against current HEAD c14517a. The legacy dense path derives its FIA geometry only from `cum_query_lens_cpu`/`seq_lens_cpu` via `_sfa_fia_dense_prefill_lengths`; `_validate_sfa_fia_shared_group` and `_execute_sfa_fia_shared_group` then pass those derived Python length lists directly as `actual_seq_lengths` / `actual_seq_lengths_kv`. The inherited device-side `cum_query_lens`/`seq_lens` are not read before this temporary dense metadata object dies. I’m therefore leaving them untouched rather than adding assignments that do not affect this call path.

File: vllm_ascend/attention/sfa_v1.py [hunk 3]
```diff
@@ -1124,6 +1202,515 @@ def _execute_sparse_flash_attention_process(
...
+        dense_output = self._execute_sfa_fia_shared_group(dense_q, dense_rope, kv_cache, dense_metadata)
+        assert dense_output is not None, "Validated shared FIA group unexpectedly declined"
+
+        tail_q = pack_rows(ql_nope, tail_spans)
+        tail_rope = pack_rows(q_pe, tail_spans)
+        tail_indices = pack_rows(topk_indices, tail_spans)
+        tail_metadata = copy(attn_metadata)
+        tail_metadata.block_table = attn_metadata.block_table[list(tail_requests)]
+        tail_metadata.cum_query_lens = cum_query_lens.new_tensor(list(tail_query_ends))
+        tail_metadata.seq_lens = seq_lens.new_tensor(list(tail_kv_lengths))
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply vllm_ascend/attention/sfa_v1.py:1910]: Validated against current HEAD c14517a. As in the multi-FIA tail case, this legacy tail path passes the rebuilt request-local `cum_query_lens`/`seq_lens` tensors explicitly into `_execute_sparse_flash_attention_process`; the downstream sparse operator consumes those explicit lengths and block table, not the CPU mirror fields. The temporary copied metadata is not reused after the tail call/row restoration. No code change is needed for this path.

File: tests/e2e/pull_request/one_card/test_sfa_v1_shared_prefill.py [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+def _main_benchmark() -> None:
+    parser = argparse.ArgumentParser(description="Run minimal native shared-prefill SFA benchmark")
+    parser.add_argument("--iters", type=int, default=16)
+    parser.add_argument("--warmup", type=int, default=8)
+    args = parser.parse_args()
+
+    cases = {
+        "single_q4096": ((4096,), (4096,), 2026),
+        "group_q1536_kv2560": ((1536, 1536), (2560, 2560), 2027),
+    }
```
[the remark this answers is not in the document]
Comment (Ruiqiu-Zheng) [reply tests/e2e/pull_request/one_card/test_sfa_v1_shared_prefill.py:555]: Addressed on current HEAD `4a426c0a0efa518a2e4dad2e2f5e7c6a7e3f3cf7`. The benchmark now includes `generic_heterogeneous_q3007_4111_1074` with `AscendAttentionState.PrefillNoCache`, in addition to the existing cache-hit cases. This directly exercises the heterogeneous request-local Multi-FIA benchmark path requested here; no further code change is needed for this thread.

**Comment by Ruiqiu-Zheng** [comment]
CI triage on current HEAD `4a426c0a0efa518a2e4dad2e2f5e7c6a7e3f3cf7`: pre-commit and `cpu-ut / cpu-ut card-` completed successfully (including `Run selected tests without device`), while the selected E2E jobs were skipped and only `ci-gate` failed. The parent #16091 is running selected E2E under `ready-precise`. Could a maintainer add `ready-precise` here to trigger the selected E2E coverage? Current evidence does not indicate a PR-owned test/code failure.

**Comment by Ruiqiu-Zheng** [comment]
Current HEAD `631cea9c979a8d38de7b8b85eb4f79d2e1886c0c` is the final generic Multi-FIA follow-up directly on #16091. Pre-commit, CPU UT, DCO, main, docs, and PR checks are green; the only failing check is ci-gate, which explicitly requests a maintainer test label for source changes. Could a maintainer add `ready-precise` to run the selected precision/E2E coverage? The former exact 3x2267 case from #16267 is retained as regression coverage, while the production path is generic. Thanks.
